keps/sig-instrumentation/20190605-metrics-validation-and-verification.md

```diff
@@ -0,0 +1,180 @@
+
+Second step will be extracting information from stable metric definitions. This step will be expecting exact structure of AST subtree of new metric call matching format proposed above. If metric options object was initialized outside of new metric call or call structure deviates from expected format then static analysis should fail. This restriction will ensure that no field was read incorrectly or stable metric missed.
+
+
+### Format of stable metrics list file
+
+This file should include information about metrics needed to detect a breaking change. Those fields should include `name`, `namespace`, `subsystem` as they are used to build metric fully-qualified name, metric type and metric labels. Metric labels should include both vector labels and const labels. 
+
+Metric information stored in file should be in easly readable and reviewable format. For that pretty printed json will be used. Example:
+```json
+[
+  {
+    "name": "some_metric",
+    "subsystem": "some_subsystem",
+    "namespace": "some_namespace",
+    "labels": [
+      "some-label"
+    ],
+    "type": "Counter"
+  }
+]
+```
+
+Json keys and lists will be sorted alphabetically to ensure that there is only one correct output.
+
+### Failure Modes
+
+In cases where static analysis could not be done correctly (invalid stable metric format, etc.) script should fail (non zero response code) and write to STDERR information on problem. Example:
+```
+pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:42 stable metric CounterOpts should be invoked from newCounter
+pkg/controller/volume/persistentvolume/scheduler_bind_cache_metrics.go:44 stable metric CounterOpts should be invoked from newCounter
+exit status 1
+```
+ 
+### Performance
+ 
+Proposed static analysis takes around 4 seconds on one CPU core. This time is comparable to conformance tests check ~2 sec and should be acceptable.
+
+## Graduation Criteria
+
+- [ ] Stable metric validation is enabled as pre-submit check
+
+## Implementation History
+
+- [ ] Implement static analysis script analysis all metrics types
+ 
+## Test Plan
+ 
+- [ ] Static analysis is covered by unit tests using common patterns of metric definition
+ 
+## Unresolved questions
+ 
+### Should changes in histogram buckets/summary objectives be a breaking change?
+ 
+Histogram buckets are usually used to calculate percentile and not directly used. This means that adjusting of buckets will not directly break their users as they should not hardcode bucket values.
+Story can be somewhat reversed for Summary objectives. Users will be directly picking set of preferred quantiles (p99) and hardcoding their values in queries.
```

Comment:
I don't understand this comment. What does "story can be somewhat reversed for Summary" mean? That we do need to include stability guarantees for Summary bucketing?

Reply:
Agreed

Reply:
It would be useful to discuss what we want to treat as breaking for purpose of this KEP.
Based on @logicalhan other comment 
`Label value changes should be ingestion compatible` options like Summary objective would not be breaking as it would only change only label values. Still in my opinion it would impact user as described.
@brancz WHYT? 
Deciding on which fields are breaking could be postponed if we decided to store all metadata in `stable-metrics-list.yaml`. It would be up to reviewer to decide.

Reply:
Agreed. I was actually thinking this further and I think it could be argued that a new label could actually be backward compatible. Actually all aggregations would continue to work, only where single time-series are expected there could be multiple, which in a vector based system is an odd expectation so I'm thinking adding a new label would actually be ok.

Reply:
I went with approach of leaving it up to reviewer to decide if change to metric is backward compatible. This requires adding all metric metadata, but as result we will get stretch goal proposed by @ehashman of generating source of metric documentation. 

@logicalhan WDYT? We wanted only to monitor backward incompatible changes but it will be hard to imminently pin down which fields we should include.

Reply:
Adding new labels is not backwards compatible because depending on your ingestion pipeline, it will not work out of the box. Adding new label values will work because the schema of your metric does not change.

Reply:
Just because new labels are possibly backwards compatible doesn't mean that they are backwards compatible. We should only be considering worst cases here, since we are looking to avoid breakages.

Reply:
I'm saying let's not categorically say no, I think we should decide when we actually have such a case and can analyze its effect appropriately.